<a href="https://colab.research.google.com/github/MykolaRobot/ML_homework_1/blob/master/WeatherAUS.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [5]:
from google.colab import files

uploaded = files.upload()


Saving knn.py to knn.py


In [0]:
import pandas as pd
import knn
import data_preprocessing


In [0]:
df=pd.read_csv('weatherAUS.csv')

In [44]:
# shape of data
df.shape

(140493, 14)

The target columns is 'RainTomorrow'.

In [0]:
# columns types
df.dtypes

Unnamed: 0         int64
MinTemp          float64
MaxTemp          float64
Rainfall         float64
WindGustSpeed    float64
WindSpeed9am     float64
WindSpeed3pm     float64
Humidity9am      float64
Humidity3pm      float64
Pressure9am      float64
Pressure3pm      float64
Temp9am          float64
Temp3pm          float64
RainToday        float64
RISK_MM          float64
RainTomorrow       int64
dtype: object

In [0]:
# get stat info about numerical columns
df.describe()

MinTemp        MaxTemp       Rainfall  WindGustSpeed  \
count  139962.000000  140232.000000  139243.000000  132424.000000   
mean       12.172039      23.211285       2.340711      39.993725   
std         6.408711       7.129024       8.387726      13.589596   
min        -8.500000      -4.800000       0.000000       6.000000   
25%         7.600000      17.900000       0.000000      31.000000   
50%        12.000000      22.600000       0.000000      39.000000   
75%        16.800000      28.200000       0.800000      48.000000   
max        33.900000      48.100000     367.600000     135.000000   

        WindSpeed9am   WindSpeed3pm    Humidity9am    Humidity3pm  \
count  139532.000000  138277.000000  138929.000000  137134.000000   
mean       14.045882      18.685450      68.821528      51.441240   
std         8.873230       8.773953      19.075249      20.807944   
min         0.000000       0.000000       0.000000       0.000000   
25%         7.000000      13.000000      57.000000      36.000000   
50%        13.000000      19.000000      70.000000      52.000000   
75%        19.000000      24.000000      83.000000      66.000000   
max       130.000000      87.000000     100.000000     100.000000   

         Pressure9am    Pressure3pm        Temp9am        Temp3pm  \
count  127666.000000  127708.000000  139669.000000  137883.000000   
mean     1017.658367    1015.263870      16.969569      21.672260   
std         7.107772       7.038742       6.502860       6.948344   
min       980.500000     977.100000      -7.200000      -5.400000   
25%      1013.000000    1010.500000      12.200000      16.600000   
50%      1017.600000    1015.200000      16.700000      21.100000   
75%      1022.400000    1020.000000      21.500000      26.400000   
max      1041.000000    1039.600000      40.200000      46.700000   

           RainToday   RainTomorrow  
count  139243.000000  140493.000000  
mean        0.223250       0.223805  
std         0.416426       0.416794  
min         0.000000       0.000000  
25%         0.000000       0.000000  
50%         0.000000       0.000000  
75%         0.000000       0.000000  
max         1.000000       1.000000

In [0]:
# count how many non-null valeus each column has
df.count().sort_values()

Pressure9am      127666
Pressure3pm      127708
WindGustSpeed    132424
Humidity3pm      137134
Temp3pm          137883
WindSpeed3pm     138277
Humidity9am      138929
Rainfall         139243
RainToday        139243
WindSpeed9am     139532
Temp9am          139669
MinTemp          139962
MaxTemp          140232
RainTomorrow     140493
dtype: int64

In [0]:
# explore columns 'RainToday' and 'RainTomorrow'. They have two unique values: 'Yes' and 'No'.
df['RainToday'].value_counts()

0.0    108157
1.0     31086
Name: RainToday, dtype: int64

In [0]:
df['RainTomorrow'].value_counts()

0    109050
1     31443
Name: RainTomorrow, dtype: int64

# Strategy 1: drop rows with nans

In [0]:
df_1=data_preprocessing.delete_nan_rows(df)
df_1.shape

In [0]:
df_1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 119590 entries, 0 to 140492
Data columns (total 14 columns):
MinTemp          119590 non-null float64
MaxTemp          119590 non-null float64
Rainfall         119590 non-null float64
WindGustSpeed    119590 non-null float64
WindSpeed9am     119590 non-null float64
WindSpeed3pm     119590 non-null float64
Humidity9am      119590 non-null float64
Humidity3pm      119590 non-null float64
Pressure9am      119590 non-null float64
Pressure3pm      119590 non-null float64
Temp9am          119590 non-null float64
Temp3pm          119590 non-null float64
RainToday        119590 non-null float64
RainTomorrow     119590 non-null int64
dtypes: float64(13), int64(1)
memory usage: 13.7 MB


use LogisticRegression as algorithm

In [0]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split

In [0]:
def workflow(df):
  X=df[df.columns[:-1]]
  y=df['RainTomorrow']
  X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3,random_state=42)
  lr=LogisticRegression().fit(X_train,y_train)
  y_pred=lr.predict(X_test)
  return y_test,y_pred

In [0]:
# calculate accuraccy
def get_accuracy(y_true,y_pred):
    matches=0
    for i in range(len(y_true)):
        if y_true[i]==y_pred[i]:
            matches+=1
    return matches/len(y_true)*100

In [0]:
y_test,y_pred=workflow(df_1)

/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


In [0]:
score=get_accuracy(list(y_test),y_pred)
score

85.00989491875018


First strategy - 85.03 % accuracy

# Strategy 2. Change misiing values to mean

In [0]:
df_2=data_preprocessing.replace_null_values(df,replace_to='mean')

In [0]:
df_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 140493 entries, 0 to 140492
Data columns (total 14 columns):
MinTemp          140493 non-null float64
MaxTemp          140493 non-null float64
Rainfall         140493 non-null float64
WindGustSpeed    140493 non-null float64
WindSpeed9am     140493 non-null float64
WindSpeed3pm     140493 non-null float64
Humidity9am      140493 non-null float64
Humidity3pm      140493 non-null float64
Pressure9am      140493 non-null float64
Pressure3pm      140493 non-null float64
Temp9am          140493 non-null float64
Temp3pm          140493 non-null float64
RainToday        140493 non-null float64
RainTomorrow     140493 non-null float64
dtypes: float64(14)
memory usage: 15.0 MB


In [0]:
y_test,y_pred=workflow(df_2)

/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


In [0]:
score=get_accuracy(list(y_test),y_pred)
score

83.44642687672012

Second strategy - 83.44 % accuracy

# Strategy 3. Replace by linear regression and use scaling to one scale

In [0]:

# all columns has missing values, that's why we need to clean some of them and use them as train_data to linear regression
df=data_preprocessing.replace_null_values(df,col=['MinTemp','MaxTemp'],replace_to='mean')

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 140493 entries, 0 to 140492
Data columns (total 14 columns):
MinTemp          140493 non-null float64
MaxTemp          140493 non-null float64
Rainfall         139243 non-null float64
WindGustSpeed    132424 non-null float64
WindSpeed9am     139532 non-null float64
WindSpeed3pm     138277 non-null float64
Humidity9am      138929 non-null float64
Humidity3pm      137134 non-null float64
Pressure9am      127666 non-null float64
Pressure3pm      127708 non-null float64
Temp9am          139669 non-null float64
Temp3pm          137883 non-null float64
RainToday        139243 non-null float64
RainTomorrow     140493 non-null float64
dtypes: float64(14)
memory usage: 15.0 MB


In [35]:
df = data_preprocessing.replace_null_by_linear_regression(df, 'Rainfall', col_based_on=['MinTemp', 'MaxTemp'])

/content/data_preprocessing.py:106: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  df_pred[target_column] = pred_values


In [0]:
columns=['WindGustSpeed','WindSpeed9am','WindSpeed3pm','Humidity9am','Humidity3pm','Pressure9am','Pressure3pm','Temp9am','Temp3pm','RainToday']

In [0]:
for col in columns:
  df=data_preprocessing.replace_null_by_linear_regression(df, col, col_based_on=['MinTemp', 'MaxTemp','Rainfall'])

In [38]:
# check 
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 140493 entries, 13860 to 140492
Data columns (total 14 columns):
MinTemp          140493 non-null float64
MaxTemp          140493 non-null float64
Rainfall         140493 non-null float64
WindGustSpeed    140493 non-null float64
WindSpeed9am     140493 non-null float64
WindSpeed3pm     140493 non-null float64
Humidity9am      140493 non-null float64
Humidity3pm      140493 non-null float64
Pressure9am      140493 non-null float64
Pressure3pm      140493 non-null float64
Temp9am          140493 non-null float64
Temp3pm          140493 non-null float64
RainToday        140493 non-null float64
RainTomorrow     140493 non-null float64
dtypes: float64(14)
memory usage: 16.1 MB


In [29]:
df.head()

MinTemp       MaxTemp      Rainfall  WindGustSpeed  WindSpeed9am  \
13860      0.0  4.988103e-16 -5.316570e-17       0.000000 -2.008238e-16   
14000      0.0  3.213416e-01 -1.062915e-01      -0.045707 -8.112059e-02   
14187      0.0  7.285092e-01 -2.409720e-01      -0.103623 -1.839074e-01   
14292      0.0 -3.806715e-01  1.259163e-01       0.054147  9.609803e-02   
14427      0.0  1.627086e+00 -5.381979e-01      -0.231436 -4.107472e-01   

       WindSpeed3pm   Humidity9am   Humidity3pm  Pressure9am   Pressure3pm  \
13860      0.000000 -7.466634e-16 -1.027686e-15     0.000000 -3.335050e-14   
14000     -0.055588 -2.339922e-01 -3.595551e-01    -0.002581 -6.576839e-02   
14187     -0.126024 -5.304806e-01 -8.151424e-01    -0.005851 -1.491026e-01   
14292      0.065852  2.771947e-01  4.259404e-01     0.003058  7.791134e-02   
14427     -0.281467 -1.184800e+00 -1.820577e+00    -0.013069 -3.330127e-01   

        Temp9am   Temp3pm  RainToday  RainTomorrow  
13860  0.000000  0.000000   0.000000      1.862305  
14000  0.157002  0.322997  -0.189250      1.862305  
14187  0.355937  0.732262  -0.429045     -0.536969  
14292 -0.185990 -0.382632   0.224191      1.862305  
14427  0.794967  1.635467  -0.958249      1.862305

In [40]:
X=df[df.columns[:-1]]
# scaling data using standart scaler -  (x[i]-mean)/std
X=data_preprocessing.standart_scaler(X)
y=df['RainTomorrow']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3,random_state=42)
lr=LogisticRegression().fit(X_train,y_train)
y_pred=lr.predict(X_test)


/content/data_preprocessing.py:79: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  df[column] = (df[column] - mean) / std
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


In [41]:
X.head()

MinTemp       MaxTemp      Rainfall  WindGustSpeed  WindSpeed9am  \
13860      0.0  4.988103e-16 -5.316570e-17       0.000000 -2.008238e-16   
14000      0.0  3.213416e-01 -1.062915e-01      -0.045707 -8.112059e-02   
14187      0.0  7.285092e-01 -2.409720e-01      -0.103623 -1.839074e-01   
14292      0.0 -3.806715e-01  1.259163e-01       0.054147  9.609803e-02   
14427      0.0  1.627086e+00 -5.381979e-01      -0.231436 -4.107472e-01   

       WindSpeed3pm   Humidity9am   Humidity3pm  Pressure9am   Pressure3pm  \
13860      0.000000 -7.466634e-16 -1.027686e-15     0.000000 -3.335050e-14   
14000     -0.055588 -2.339922e-01 -3.595551e-01    -0.002581 -6.576839e-02   
14187     -0.126024 -5.304806e-01 -8.151424e-01    -0.005851 -1.491026e-01   
14292      0.065852  2.771947e-01  4.259404e-01     0.003058  7.791134e-02   
14427     -0.281467 -1.184800e+00 -1.820577e+00    -0.013069 -3.330127e-01   

        Temp9am   Temp3pm  RainToday  
13860  0.000000  0.000000   0.000000  
14000  0.157002  0.322997  -0.189250  
14187  0.355937  0.732262  -0.429045  
14292 -0.185990 -0.382632   0.224191  
14427  0.794967  1.635467  -0.958249

In [42]:
score=get_accuracy(list(y_test),y_pred)
score

84.08939925975135

Third strategy - 84.089 % accuracy

# Strategy 4: replace by knn and scale using min max scaler

In [0]:

# all columns has missing values, that's why we need to clean some of them and use them as train_data to knn
df=data_preprocessing.replace_null_values(df,col=['MinTemp','MaxTemp'],replace_to='mean')

In [0]:
columns=['Rainfall','WindGustSpeed','WindSpeed9am','WindSpeed3pm','Humidity9am','Humidity3pm','Pressure9am','Pressure3pm','Temp9am','Temp3pm','RainToday']
for column in columns:
  df=knn.knn(df,column,k=5)

In [50]:
# check knn
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 140493 entries, 0 to 140492
Data columns (total 14 columns):
MinTemp          140493 non-null float64
MaxTemp          140493 non-null float64
Rainfall         140493 non-null float64
WindGustSpeed    140493 non-null float64
WindSpeed9am     140493 non-null float64
WindSpeed3pm     140493 non-null float64
Humidity9am      140493 non-null float64
Humidity3pm      140493 non-null float64
Pressure9am      140493 non-null float64
Pressure3pm      140493 non-null float64
Temp9am          140493 non-null float64
Temp3pm          140493 non-null float64
RainToday        140493 non-null float64
RainTomorrow     140493 non-null float64
dtypes: float64(14)
memory usage: 15.0 MB


In [0]:
# use min_max scaler
X=df[df.columns[:-1]]
# scaling data using standart scaler -  (x[i]-mean)/std
X=data_preprocessing.min_max_scaler(X)
y=df['RainTomorrow']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3,random_state=42)
lr=LogisticRegression().fit(X_train,y_train)
y_pred=lr.predict(X_test)


In [52]:
score=get_accuracy(list(y_test),y_pred)
score

83.68131346683117

strategy 4: accuracy - 83.681


# Conclusion

1. 4 strategies were tried.
2. Best one is to drop rows with missing values and it has accuracy around 85%.
3. Other strategies give accuracy around 83-84%.
4. My methods work much slower than pandas functions:)
That's all!